routing-ortools/main.py

```diff
@@ -32,46 +36,308 @@ def main():
+    lats_destination = np.tile(lats_destination, len(lats_destination))
+    lons_destination = np.tile(lons_destination, len(lons_destination))
+
+    distances = haversine(
+        lats_origin=lats_origin,
+        lons_origin=lons_origin,
+        lats_destination=lats_destination,
+        lons_destination=lons_destination,
+    )
+
+    # Convert the distances to a square matrix.
+    num_locations = len(input_data["stops"]) + 2 * len(input_data["vehicles"])
+    matrix = distances.reshape(num_locations, num_locations)
+
+    end = time.time()
+    log(f"Distance matrix calculation took {round(end - start, 2)} seconds.")
+    return matrix
+
+
+def process_distance_matrix(input_data: dict[str, Any]) -> None:
+    """Calculates the distance matrix for the input data."""
+
+    # If the input data already contains a distance matrix, return it.
+    if "distance_matrix" in input_data:
+        np_matrix = np.array(input_data["distance_matrix"])
+        input_data["distance_matrix"] = _expand_missing_start_end(np_matrix, input_data)
+
+    # Only calculate the distance matrix if there is no duration matrix.
+    if "duration_matrix" not in input_data and "distance_matrix" not in input_data:
+        input_data["distance_matrix"] = _calculate_distance_matrix(input_data)
+
+    # Make sure the matrix is integer (round the values).
+    if "distance_matrix" in input_data:
+        input_data["distance_matrix"] = np.rint(input_data["distance_matrix"]).astype(int)
+
+
+def process_duration_matrix(input_data: dict[str, Any]) -> None:
+    """Prepares the duration matrix of the input data, if given."""
+
+    # If the input data already contains a duration matrix, return it.
+    if "duration_matrix" in input_data:
+        np_matrix = np.array(input_data["duration_matrix"])
+        input_data["duration_matrix"] = _expand_missing_start_end(np_matrix, input_data)
+
+    # Make sure the matrix is integer (round the values).
+    if "duration_matrix" in input_data:
+        input_data["duration_matrix"] = np.rint(input_data["duration_matrix"]).astype(int)
+
+
+def haversine(
+    lats_origin: np.ndarray | float,
+    lons_origin: np.ndarray | float,
+    lats_destination: np.ndarray | float,
+    lons_destination: np.ndarray | float,
+) -> np.ndarray | float:
+    """Calculates the haversine distance between arrays of coordinates."""
+
+    lons_destination, lats_destination, lons_origin, lats_origin = map(
+        np.radians,
+        [lons_destination, lats_destination, lons_origin, lats_origin],
+    )
+    delta_lon = lons_destination - lons_origin
+    delta_lat = lats_destination - lats_origin
+    term1 = np.sin(delta_lat / 2.0) ** 2
+    term2 = np.cos(lats_origin) * np.cos(lats_destination) * np.sin(delta_lon / 2.0) ** 2
+    a = term1 + term2
+    c = 2 * np.arcsin(np.sqrt(a))
+    earth_radius = 6371000
+
+    return earth_radius * c
+
+
 def solve(input_data: dict[str, Any], duration: int) -> dict[str, Any]:
```

Comment:
I think the `solve` func should come right after `main`, to be consistent with the other apps.